pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,375 @@
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ClusterIngress.
+	Items []ClusterIngress `json:"items"`
+}
+
+// ClusterIngressSpec describes the ClusterIngress the user wishes to exist.
+type ClusterIngressSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// TLS configuration. Currently the ClusterIngress only supports a single TLS
+	// port, 443. If multiple members of this list specify different hosts, they
+	// will be multiplexed on the same port according to the hostname specified
+	// through the SNI TLS extension, if the ingress controller fulfilling the
+	// ingress supports SNI.
+	// +optional
+	TLS []ClusterIngressTLS `json:"tls,omitempty"`
+
+	// A list of host rules used to configure the ClusterIngress. If unspecified, or
+	// no rule matches, all traffic is sent to the default backend.
+	// +optional
+	Rules []ClusterIngressRule `json:"rules,omitempty"`
+}
+
+// ClusterIngressTLS describes the transport layer security associated with an ClusterIngress.
+type ClusterIngressTLS struct {
+	// Hosts are a list of hosts included in the TLS certificate. The values in
+	// this list must match the name/s used in the tlsSecret. Defaults to the
+	// wildcard host setting for the loadbalancer controller fulfilling this
+	// ClusterIngress, if left unspecified.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+	// SecretName is the name of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
+	// +optional
+	SecretName string `json:"secretName,omitempty"`
+
+	// SecretNamespace is the namespace of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
+	// +optional
+	SecretNamespace string `json:"secretNamespace,omitempty"`
+
+	// ServerCertificate identifies the certificate filename in the secret.
+	// Defaults to `tls.cert`.
+	// +optional
+	ServerCertificate string `json:"serverCertificate,omitempty"`
+
+	// PrivateKey identifies the private key filename in the secret.
+	// Defaults to `tls.key`.
+	// +optional
+	PrivateKey string `json:"privateKey,omitempty"`
+}
+
+// ConditionType represents a ClusterIngress condition value
+const (
+	// ClusterIngressConditionReady is set when the clusterIngress networking setting is
+	// configured and it has a load balancer address.
+	ClusterIngressConditionReady = duckv1alpha1.ConditionReady
+
+	// ClusterIngressConditionNetworkConfigured is set when the ClusterIngress's underlying
+	// network programming has been configured.
+	ClusterIngressConditionNetworkConfigured duckv1alpha1.ConditionType = "NetworkConfigured"
```

Comment:
So I think it is considered to reflect the status of underlying networking model objects(virtualservice now), but not the status of workload service?
And when the target revision is not ready, but the underlying virtualservice object that redirects to avtivator has been created, the condition here should also be set to true, right?

Reply:
That is correct.  I've updated the doc to make it clear.

Reply:
I don't know if you want to echo back the configured network state (ala Route). This can be helpful if there are certain properties on `HTTPClusterIngressRuleValue` which can't be represented by some implementations. (Though maybe you just want to avoid that case and declare those insufficient implementations.)